internal/api/middleware.go

```diff
@@ -298,31 +299,37 @@ func (t *timeoutResponseWriter) WriteHeader(statusCode int) {
 	t.Lock()
 	defer t.Unlock()
 
+	if t.wroteHeader {
+		// ignore multiple calls to WriteHeader
+		// once WriteHeader has been called once, a snapshot of the header map is taken
+		// and saved in snapHeader to be used in finallyWrite
+		return
+	}
 	t.statusCode = statusCode
 	t.wroteHeader = true
+	t.snapHeader = t.header.Clone()
 }
 
 func (t *timeoutResponseWriter) finallyWrite(w http.ResponseWriter) {
 	t.Lock()
 	defer t.Unlock()
 
 	dst := w.Header()
-	for k, vv := range t.header {
+	for k, vv := range t.snapHeader {
 		dst[k] = vv
 	}
+
 	if !t.wroteHeader {
 		t.statusCode = http.StatusOK
 	}
 
-	fmt.Printf("@@@@@@@@@@@@@@@@@@@@@@ %v %v\n", t.header, dst)
-
 	w.WriteHeader(t.statusCode)
 	if _, err := w.Write(t.buf.Bytes()); err != nil {
 		logrus.WithError(err).Warn("Write failed")
 	}
 }
 
-func (a *API) timeoutMiddleware(timeout time.Duration) func(http.Handler) http.Handler {
+func timeoutMiddleware(timeout time.Duration) func(http.Handler) http.Handler {
```

Comment:
In favour of  decoupling as it looks we don't have or no longer have any dependencies on API state